### PR TITLE
feat: support co-partitioned range right-side equi hash joins

### LINKED ISSUE #23453: Allow co-partitioned `Partitioning::Range` inputs for right-side hash joins
Related:

- [#22395](https://github.com/apache/datafusion/issues/22395)
- [#21992](https://github.com/apache/datafusion/issues/21992)
- [#23183](https://github.com/apache/datafusion/issues/23183)
- [#23184](https://github.com/apache/datafusion/pull/23184)
- [#23266](https://github.com/apache/datafusion/issues/23266)

## Is your feature request related to a problem or challenge?

#23184 lets compatible range-partitioned inputs satisfy inner partitioned hash joins without repartitioning. Right-side partitioned equi joins still use the repartition path even when both inputs are already co-partitioned by the join keys.

## Example

```text
HashJoinExec: mode=Partitioned, join_type=Right, on=[(range_key@0, range_key@0)]
  DataSourceExec: output_partitioning=Range([range_key@0 ASC], [(10), (20), (30)], 4)
  DataSourceExec: output_partitioning=Range([range_key@0 ASC], [(10), (20), (30)], 4)
```

The compatible range inputs are already co-partitioned by partition index. The optimizer should preserve that layout instead of inserting hash repartitions for the right join.

## Describe the solution you'd like

Extend `InputDistributionRequirements::co_partitioned(...)` to right-side partitioned equi hash joins where execution remains partition-local.

Covered join types:

- `Right`
- `RightSemi`
- `RightAnti`
- `RightMark`

## Acceptance criteria

- Compatible range/range inputs avoid hash repartitioning for the covered join types.
- Incompatible split points, partition counts, sort options, or key expressions still trigger repartitioning.
- Tests cover matched and unmatched rows.
- `RightMark` tests cover null-related marker behavior.
- Partitioned dynamic filters remain unchanged unless separately addressed by #23376.

## Additional context

### PR DESCRIPTION
**Pull request by gmhelmold**
## Which issue does this PR close?

- Closes #23453.
- Part of #22395.

## Rationale for this change

#23184 let compatible range-partitioned inputs satisfy **inner** partitioned hash joins without repartitioning. Right-side partitioned equi joins have the same locality guarantee: `Right`, `RightSemi`, `RightAnti` and `RightMark` all anchor every output row on the probe (right) partition (`on_lr_is_preserved` is probe-side for all four), so when both inputs are co-partitioned by the join keys, execution stays partition-local and the hash repartition is unnecessary.

This removes unnecessary `RepartitionExec`s for already-co-located inputs, extending the inner-join behavior from #23184 to the right-side variants. No micro-benchmark included, consistent with #23184; correctness is demonstrated by matched/unmatched execution results below.

## What changes are included in this PR?

The only production change is widening the existing inner-only gate in `HashJoinExec::input_distribution_requirements` from `join_type == JoinType::Inner` to `matches!(join_type, Inner | Right | RightSemi | RightAnti | RightMark)` (under `PartitionMode::Partitioned`). The `co_partitioned` / range-satisfaction machinery from #23184 is unchanged — the sanity checker and enforce_distribution consume range satisfaction join-type-agnostically.

One behavior note for #23376: range co-partitioned right joins now stay `Range`/`Range`, so partitioned dynamic filters are disabled for them (`has_partitioned_dynamic_filter_routing` returns false), the same safe delta #23184 introduced for inner joins. These variants are probe-not-preserved for pruning, so dynamic filters were not eligible to prune their probe rows regardless.

## Are these changes tested?

Yes.

- Optimizer (`enforce_distribution.rs`): 4 reuse tests (one per join type) proving compatible range/range inputs keep `Range` partitioning with no `RepartitionExec`; 4 incompatibility tests proving that mismatched split points, sort options, partition counts, or join-key expressions still insert a hash repartition; sanity-check pairs mirroring #23184.
- Execution (`range_partitioning.slt`): `EXPLAIN` plan pins plus matched/unmatched result checks for `Right` (incl. `NULL` left values), `RightSemi`, `RightAnti`, and an incompatible-layout `Right` join (repartitions, correct results).
- The new reuse tests fail on `main` without the gate change (verified by reverting the production diff: exactly the 4 reuse tests fail, everything else passes).
- `RightMark` testing note: `RightMark` is not reachable from SQL in sqllogictest (`IN`-subquery decorrelation emits `LeftMark`; physical `RightMark` only appears via a statistics-based swap), so its co-partitioning behavior is pinned at the optimizer/plan level, and mark null-marker semantics (matched/unmatched/`NULL` build keys) are pinned via the `LeftMark` path in the slt.

## Are there any user-facing changes?

No API changes. Plans over compatible range-partitioned inputs avoid a hash repartition for right-side equi hash joins.

### CODE REVIEW

File: datafusion/core/tests/physical_optimizer/enforce_distribution.rs [hunk 1]
```diff
@@ -870,6 +870,298 @@ fn range_inner_hash_join_rehashes_incompatible_range_partitioning() -> Result<()
     Ok(())
 }
 
+#[test]
+fn range_right_hash_join_reuses_range_partitioning() -> Result<()> {
```
Comment (gene-bordegaray) [inline datafusion/core/tests/physical_optimizer/enforce_distribution.rs]: note before I do full review: we are doing most end to end testing in `range_partitioning.slt` rather than the unit to ensure behavior works across all distribution rules. Our split has been 1-2 unit tests for smoke or hard to replicate scenarios in slt, then the bulk in the slt test. Thank you 😄

Comment (gmhelmold) [reply datafusion/core/tests/physical_optimizer/enforce_distribution.rs]: Makes sense — restructuring now: moving the end-to-end matrix into `range_partitioning.slt` and keeping a couple of unit smoke tests in `enforce_distribution.rs` for the hard-to-replicate boundary cases. Thanks for the pointer! 😄

**Comment by gmhelmold** [comment]
Done in 852a619: the end-to-end matrix now lives in `range_partitioning.slt` (reuse cases for Right/RightSemi/RightAnti + rehash cases for split-points, partition-count, and non-range join keys — the last two needed one small narrow-fixture addition to the harness). Kept two unit tests in `enforce_distribution.rs` with comments explaining why they can't move: `RightMark` plans aren't SQL-producible, and the sort-direction axis needs a descending Range input the SLT fixtures don't declare. Net: unit block 8→2, no scenario lost.

File: datafusion/sqllogictest/src/test_context/range_partitioning.rs [hunk 2]
```diff
@@ -93,6 +93,34 @@ pub(super) fn register_range_partitioned_table(ctx: &SessionContext) {
         ],
         Some(shifted_output_partitioning),
     );
+
+    // Same rows as `range_partitioned` but split into only three range
+    // partitions on `range_key`. Used to exercise the co-partition check when
+    // two Range inputs disagree on partition count.
+    let narrow_output_partitioning = Partitioning::Range(
```
Comment (gene-bordegaray) [inline datafusion/sqllogictest/src/test_context/range_partitioning.rs:137]: I like it 👍

File: datafusion/sqllogictest/test_files/range_partitioning.slt [hunk 3]
```diff
@@ -634,6 +635,260 @@ ORDER BY l.range_key;
...
+50 5 50
+100 10 100
+150 15 150
+NULL 20 200
+NULL 25 250
+NULL 30 300
+NULL 35 350
+
+##########
+# TEST 22: Mark Join Marker Semantics over Range Partitioned Inputs
```
Comment (gene-bordegaray) [inline datafusion/sqllogictest/test_files/range_partitioning.slt]: nit: can we move this as the last test in this block?

File: datafusion/sqllogictest/test_files/range_partitioning.slt [hunk 4]
```diff
@@ -634,6 +635,260 @@ ORDER BY l.range_key;
...
+query TT
+EXPLAIN SELECT r.range_key, r.value
+FROM range_partitioned r
+WHERE r.non_range_key = 2 OR r.range_key IN (
+  SELECT range_key FROM range_partitioned WHERE value <= 150);
+----
+physical_plan
+01)FilterExec: non_range_key@1 = 2 OR mark@3, projection=[range_key@0, value@2]
+02)--HashJoinExec: mode=Partitioned, join_type=LeftMark, on=[(range_key@0, range_key@0)]
+03)----RepartitionExec: partitioning=Hash([range_key@0], 4), input_partitions=4
```
Comment (gene-bordegaray) [inline datafusion/sqllogictest/test_files/range_partitioning.slt]: to be clear this should disappear after #23453 ? Just putting here to show reationale on why the unit is good

File: datafusion/sqllogictest/test_files/range_partitioning.slt [hunk 5]
```diff
@@ -634,6 +635,260 @@ ORDER BY l.range_key;
 30 600
 35 700
 
+##########
```
Comment (gene-bordegaray) [inline datafusion/sqllogictest/test_files/range_partitioning.slt:632]: can we add one subset test for sanity check 👍

**Review by gene-bordegaray — COMMENTED** [review]
few nits but this is great, thank you. Feel free to ping on another round 😄

File: datafusion/sqllogictest/test_files/range_partitioning.slt (hunk 3 above)
Comment (gmhelmold) [reply datafusion/sqllogictest/test_files/range_partitioning.slt]: Done — moved to the end of the partitioned-hash-join block (now the last test there, before the SMJ/SHJ cases that came in with the main merge).

File: datafusion/sqllogictest/test_files/range_partitioning.slt (hunk 4 above)
Comment (gmhelmold) [reply datafusion/sqllogictest/test_files/range_partitioning.slt]: Good question — no, this one stays. This PR is the implementation of #23453 (it closes it), and the plan pinned here is already generated with the change applied: the range-satisfaction gate in `HashJoinExec::input_distribution_requirements` covers `Inner | Right | RightSemi | RightAnti | RightMark` only, and SQL `IN`-subquery decorrelation produces a `LeftMark` join — a left-anchored type intentionally outside this PR's covered set — so its inputs keep the Hash repartition path. And exactly right on the rationale: since `RightMark` is not reachable from SQL (it only appears when a statistics-based swap flips a `LeftMark`), its reuse behavior is pinned by the `enforce_distribution.rs` unit test, while this case pins the mark-marker semantics end-to-end on the SQL-reachable path. It would only change if co-partition reuse is later extended to the left-anchored variants under the #22395 umbrella — happy to file that follow-up if there's interest.

File: datafusion/sqllogictest/test_files/range_partitioning.slt (hunk 5 above)
Comment (gmhelmold) [reply datafusion/sqllogictest/test_files/range_partitioning.slt:632]: Added TEST 22: a RIGHT JOIN on the composite key `(range_key, non_range_key)` where `Range([range_key])` covers only a subset — pins that, unlike aggregates (TEST 3), subset satisfaction stays disabled for partitioned joins, so both sides Hash-repartition on the full key even with the threshold met.

**Comment by gmhelmold** [comment]
Round done! All three nits addressed (mark-semantics test moved to the end of its block; TEST 22 added as the composite-key subset sanity check; rationale for the LeftMark repartition answered inline) — plus the branch is synced with main twice over (absorbed the SMJ/SHJ co-partition tests, #23584's plan changes, and #23623's InterleaveExec cases; full suite green, 98 unit + 44 SLT). Ready for another look @gene-bordegaray 😄

File: datafusion/sqllogictest/test_files/range_partitioning.slt [hunk 6]
```diff
@@ -629,7 +630,304 @@ ORDER BY l.range_key;
...
+
+query TT
+EXPLAIN SELECT r.range_key, r.value
+FROM (SELECT range_key FROM range_partitioned WHERE value <= 150) l
+RIGHT ANTI JOIN range_partitioned r ON l.range_key = r.range_key;
+----
+physical_plan
+01)HashJoinExec: mode=Partitioned, join_type=RightAnti, on=[(range_key@0, range_key@0)]
+02)--FilterExec: value@1 <= 150, projection=[range_key@0]
+03)----DataSourceExec: file_groups={4 groups: [[WORKSPACE_ROOT/datafusion/sqllogictest/test_files/scratch_range_partitioning/range_partitioned/part-0.csv], [WORKSPACE_ROOT/datafusion/sqllogictest/test_files/scratch_range_partitioning/range_partitioned/part-1.csv], [WORKSPACE_ROOT/datafusion/sqllogictest/test_files/scratch_range_partitioning/range_partitioned/part-2.csv], [WORKSPACE_ROOT/datafusion/sqllogictest/test_files/scratch_range_partitioning/range_partitioned/part-3.csv]]}, projection=[range_key, value], output_partitioning=Range([range_key@0 ASC], [(10), (20), (30)], 4), file_type=csv, has_header=false
```
Comment (gene-bordegaray) [inline datafusion/sqllogictest/test_files/range_partitioning.slt]: can we ignore files 👍

File: datafusion/sqllogictest/test_files/range_partitioning.slt [hunk 7]
```diff
@@ -629,7 +630,304 @@ ORDER BY l.range_key;
 35 700
 
 ##########
-# TEST 18: Sort Merge Join Avoids Repartition for Compatible Range Inputs
+# TEST 18: Right Join on Range Partition Column
+# Right-side partitioned hash joins also opt in to Range satisfying
```
Comment (gene-bordegaray) [inline datafusion/sqllogictest/test_files/range_partitioning.slt]: lets keep descriptions generic so they remain valid after the PR merges

File: datafusion/sqllogictest/test_files/range_partitioning.slt [hunk 8]
```diff
@@ -629,7 +630,304 @@ ORDER BY l.range_key;
...
+100 10 100
+150 15 150
+NULL 20 200
+NULL 25 250
+NULL 30 300
+NULL 35 350
+
+##########
+# TEST 22: Right Join on Composite Key With Range Subset Rehashes
+# The join key (range_key, non_range_key) is a strict superset of the exposed
```
Comment (gene-bordegaray) [inline datafusion/sqllogictest/test_files/range_partitioning.slt]: coudl we keep explanations concise 👍

File: datafusion/sqllogictest/test_files/range_partitioning.slt [hunk 9]
```diff
@@ -629,7 +630,304 @@ ORDER BY l.range_key;
...
+50 5 50
+10 10 100
+50 15 150
+10 20 200
+50 25 250
+10 30 300
+50 35 350
+
+##########
+# TEST 25: Mark Join Marker Semantics over Range Partitioned Inputs
```
Comment (gene-bordegaray) [inline datafusion/sqllogictest/test_files/range_partitioning.slt]: lets just keep descriptions specific to what this test is showing. Not mentioning other things in teh codebase

**Review by gene-bordegaray — COMMENTED** [review]
sorry, trying to keep files clean and clear. Few more things 👍

**Comment by gmhelmold** [comment]
All addressed in 04e31b2 👍

- **Ignore files:** the remaining plan assertions now use `file_groups=<slt:ignore>`, so the file no longer pins absolute scratch paths (matches the rest of the file).
- **Descriptions:** trimmed TEST 18/22/25 to be concise and specific to what each test shows — no cross-references or PR-relative wording.

Verified locally: `cargo test -p datafusion-sqllogictest --test sqllogictests -- range_partitioning` passes.

**Review by gene-bordegaray — APPROVED** [review]
thanks, one last thing 👍

**Comment by gmhelmold** [comment]
Good catch that this deserves clarifying 👍 I tried the metadata-only change first, but a descending Range requires descending-ordered split points — `[10, 20, 30]` with `descending: true` fails constructor validation:

```
Range partitioning split points must be strictly ordered: split point 0 ((10)) must be less than split point 1 ((20))
```

So `[30, 20, 10]` DESC isn't a second, independent flip — it's the same partition boundaries (10/20/30) as the left's `[10, 20, 30]` ASC, just encoded in the sort's order. The only real difference between the two inputs is the sort direction, which is exactly the incompatibility under test. Added a comment above the right input in 64aba77 spelling that out for the next reader. Happy to switch to a different construction if you'd prefer.

File: datafusion/core/tests/physical_optimizer/enforce_distribution.rs [hunk 10]
```diff
@@ -870,6 +870,91 @@ fn range_inner_hash_join_rehashes_incompatible_range_partitioning() -> Result<()
     Ok(())
 }
 
+// Kept as a unit test: `RightMark` join plans cannot be produced from SQL (they
+// only arise when a statistical swap flips a `LeftMark` join around), so this
```
Comment (gene-bordegaray) [inline datafusion/core/tests/physical_optimizer/enforce_distribution.rs]: Lets try to keep very PR specific comments out of the code. 

Usually phrases like this "Kept as a unit test: ..." does not have any value outside of the scopoe of this PR becuase a new reaader will not have the ocntext that you or I have about this test. Rather lets omit the comments and allow the code to do the talking 👍

File: datafusion/core/tests/physical_optimizer/enforce_distribution.rs [hunk 11]
```diff
@@ -870,6 +870,91 @@ fn range_inner_hash_join_rehashes_incompatible_range_partitioning() -> Result<()
...
+    HashJoinExec: mode=Partitioned, join_type=RightMark, on=[(a@0, a@0)]
+      DataSourceExec: file_groups={4 groups: [[p0], [p1], [p2], [p3]]}, projection=[a, b, c, d, e], output_partitioning=Range([a@0 ASC], [(10), (20), (30)], 4), file_type=parquet
+      DataSourceExec: file_groups={4 groups: [[p0], [p1], [p2], [p3]]}, projection=[a, b, c, d, e], output_partitioning=Range([a@0 ASC], [(10), (20), (30)], 4), file_type=parquet
+    "
+    );
+
+    Ok(())
+}
+
+// Kept as a unit test: a descending Range input cannot be registered through the
```
Comment (gene-bordegaray) [inline datafusion/core/tests/physical_optimizer/enforce_distribution.rs]: ditto

File: datafusion/core/tests/physical_optimizer/enforce_distribution.rs [hunk 12]
```diff
@@ -870,6 +870,91 @@ fn range_inner_hash_join_rehashes_incompatible_range_partitioning() -> Result<()
...
+fn range_right_semi_hash_join_rehashes_incompatible_sort_options() -> Result<()> {
+    let left = parquet_exec_with_output_partitioning(range_partitioning(
+        "a",
+        [10, 20, 30],
+        SortOptions::default(),
+    )?);
+    // A descending Range requires descending-ordered split points (the constructor rejects
+    // ascending points under a descending sort), so [30, 20, 10] DESC encodes the same
+    // partition boundaries as the left's [10, 20, 30] ASC -- only the sort direction differs,
+    // which is the incompatibility under test.
```
Comment (gene-bordegaray) [inline datafusion/core/tests/physical_optimizer/enforce_distribution.rs]: ok I think we got confused here. We are specifically tring to isolate incompatible sort options based on metadata. To isolate this is would be best to use a single split point and just change the descending metadata rather than both 👍

Comment (gene-bordegaray) [reply datafusion/core/tests/physical_optimizer/enforce_distribution.rs]: this test would still fail if the sort otpions check was removed

Comment (gene-bordegaray) [reply datafusion/core/tests/physical_optimizer/enforce_distribution.rs]: And I don't think this warrants a long comments

File: datafusion/sqllogictest/test_files/range_partitioning.slt [hunk 13]
```diff
@@ -629,7 +630,298 @@ ORDER BY l.range_key;
...
+50 5 50
+100 10 100
+150 15 150
+NULL 20 200
+NULL 25 250
+NULL 30 300
+NULL 35 350
+
+##########
+# TEST 22: Right Join on Composite Key With Range Subset Rehashes
```
Comment (gene-bordegaray) [inline datafusion/sqllogictest/test_files/range_partitioning.slt]: please resplace with something like:

```    
# TEST 22: Composite-Key Right Join Repartitions
# Range([range_key]) does not satisfy a partitioned join on
# (range_key, non_range_key), so both sides repartition on the full key.
```

File: datafusion/sqllogictest/test_files/range_partitioning.slt (hunk 9 above)
Comment (gene-bordegaray) [reply datafusion/sqllogictest/test_files/range_partitioning.slt]: please replace with something like:

```text
# TEST 25: Mark Join Marker Semantics
# Mark joins preserve matched, unmatched, and NULL-key marker behavior over
# range-partitioned inputs.
```

**Review by gene-bordegaray — APPROVED** [review]
These are all non-blockinng.

Thank you for the contribution 🙇 . I kindly ask if you are using AI tools to aid coding to review the comments and output you are getting. It is important to keep comments and documentation clear in the codebase for future readers and maintainability

cc: @gabotechs for when these are fixed

**Comment by gmhelmold** [comment]
Thanks for the careful review, and fair point on the comments 🙏 You're right that phrasing like "Kept as a unit test: …" only carries meaning with the context you and I have — a future reader doesn't have it. Stripped those and let the code speak. In `82a68b9`:

- Removed the PR-specific comments on both unit tests.
- Restructured the sort-options test to a single split point with only the `descending` flag flipped, as you suggested — it's self-explanatory now, and still fails if the sort-options check is removed.
- Tightened the TEST 22 / TEST 25 descriptions to your wording.

I'll keep comments lean and reader-focused going forward — thanks for pushing on clarity.

**Comment by gene-bordegaray** [comment]
could you merge this with origin main. I want to make sure it is correct witht the recent changes in #23454 🙇

**Comment by gmhelmold** [comment]
Merged `upstream/main` (5b816da). The overlap with #23583 was the co-partition allowlist — reconciled it to the union: `Inner | Full | Right | RightSemi | RightAnti | RightMark`, so the Full-join reuse and the right-side reuse coexist. Kept both range fixtures (`narrow` + `sparse`) and renumbered the slt tests sequentially. `range_partitioning.slt` and the enforce_distribution tests pass locally.

**Comment by gene-bordegaray** [comment]
thank you, this is good now cc: @gabotechs

**Review by gabotechs — APPROVED** [review]
Thanks @gmhelmold and @gene-bordegaray!
